--- ghc-7.6.3/compiler/llvmGen/LlvmCodeGen/Base.hs~	2013-04-19 06:22:46.000000000 +0900
+++ ghc-7.6.3/compiler/llvmGen/LlvmCodeGen/Base.hs	2013-07-24 17:05:06.491900335 +0900
@@ -151,7 +151,7 @@
 minSupportLlvmVersion = 28
 
 maxSupportLlvmVersion :: LlvmVersion
-maxSupportLlvmVersion = 31
+maxSupportLlvmVersion = 33
 
 -- ----------------------------------------------------------------------------
 -- * Environment Handling
