diff -u ghc/libraries/gen_contents_index~ ghc/libraries/gen_contents_index
--- ghc-6.10.1/libraries/gen_contents_index~	2008-11-06 11:18:45.000000000 +1000
+++ ghc-6.10.1/libraries/gen_contents_index	2008-11-10 17:53:12.000000000 +1000
@@ -23,7 +23,7 @@
 
 for HADDOCK_FILE in $HADDOCK_FILES
 do
-    NAME=` echo "$HADDOCK_FILE" | sed 's#/dist/.*##' | sed 's#.*/##' `
+    NAME=` echo "$HADDOCK_FILE" | sed 's#/dist/.*##' | sed 's#.*/##' | sed 's#\.haddock##' `
     HADDOCK_ARGS="$HADDOCK_ARGS --read-interface=$NAME,$HADDOCK_FILE"
     NAMES="$NAMES $NAME"
 done
