diff -u ghc-7.4.1/compiler/main/DriverPipeline.hs.orig ghc-7.4.1/compiler/main/DriverPipeline.hs
--- ghc-7.4.1/compiler/main/DriverPipeline.hs.orig	2012-02-02 03:10:32.000000000 +0900
+++ ghc-7.4.1/compiler/main/DriverPipeline.hs	2012-04-27 10:42:53.142111769 +0900
@@ -1376,9 +1376,9 @@
         -- does not enable VFP by default. Let's do this manually here
         fpOpts = case platformArch (targetPlatform dflags) of 
                    ArchARM ARMv7 ext -> if (elem VFPv3 ext)
-                                      then ["-mattr=+v7,+vfp3"]
+                                      then ["-mattr=+v7,+vfp3", "-float-abi=hard"]
                                       else if (elem VFPv3D16 ext)
-                                           then ["-mattr=+v7,+vfp3,+d16"]
+                                           then ["-mattr=+v7,+vfp3,+d16", "-float-abi=hard"]
                                            else []
                    _               -> []
 
