diff -ur ghc-8.4.4.orig/includes/Stg.h ghc-8.4.4/includes/Stg.h
--- ghc-8.4.4.orig/includes/Stg.h	2018-01-29 21:20:01.000000000 +0000
+++ ghc-8.4.4/includes/Stg.h	2021-04-01 23:19:51.741788378 +0100
@@ -236,6 +236,8 @@
 #define EB_(X)    extern const char X[]
 #define IB_(X)    static const char X[]
 /* static (non-heap) closures (requires alignment for pointer tagging): */
+#define EI_(X)    extern       StgWordArray (X) GNU_ATTRIBUTE(aligned (8))
+#define II_(X)    static       StgWordArray (X) GNU_ATTRIBUTE(aligned (8))
 #define EC_(X)    extern       StgWordArray (X) GNU_ATTRIBUTE(aligned (8))
 #define IC_(X)    static       StgWordArray (X) GNU_ATTRIBUTE(aligned (8))
 /* writable data (does not require alignment): */
