diff -u ghc-6.12.1/ghc.mk\~ ghc-6.12.1/ghc.mk
--- ghc-6.12.1/ghc.mk~	2009-12-11 04:11:33.000000000 +1000
+++ ghc-6.12.1/ghc.mk	2010-01-09 23:17:20.000000000 +1000
@@ -846,7 +846,7 @@
 INSTALLED_GHC_PKG_REAL=$(DESTDIR)$(bindir)/ghc-pkg.exe
 endif
 
-INSTALLED_PACKAGES = $(filter-out haskeline mtl terminfo,$(PACKAGES))
+INSTALLED_PACKAGES = $(filter-out mtl,$(PACKAGES))
 HIDDEN_PACKAGES = ghc-binary
 
 define set_INSTALL_DISTDIR
