diff -u ghc-6.12.3/libraries/gen_contents_index\~ ghc-6.12.3/libraries/gen_contents_index
--- ghc-6.12.3/libraries/gen_contents_index~	2010-09-14 13:03:12.000000000 +1000
+++ ghc-6.12.3/libraries/gen_contents_index	2010-11-04 16:41:32.000000000 +1000
@@ -24,7 +24,7 @@
 *)
     HADDOCK=/usr/bin/haddock
     # We don't want the GHC API to swamp the index
-    HADDOCK_FILES=`ls -1 */*.haddock | grep -v '/ghc\.haddock' | sort`
+    HADDOCK_FILES=`ls -1 */*.haddock | grep -v '/ghc\.haddock' | grep -v '/type-level\.haddock' | sort`
     for HADDOCK_FILE in $HADDOCK_FILES
     do
         NAME_VERSION=`echo "$HADDOCK_FILE" | sed 's#/.*##'`
