--- ghc-7.7.20130828/utils/ghc-cabal/Main.hs~	2013-08-28 08:06:37.000000000 +0900
+++ ghc-7.7.20130828/utils/ghc-cabal/Main.hs	2013-09-03 17:51:22.800653817 +0900
@@ -251,8 +251,8 @@
           libsubdir = toPathTemplate "$pkgid",
           docdir    = toPathTemplate $
                           if relocatableBuild
-                          then "$topdir/../doc/html/libraries/$pkgid"
-                          else (myDocdir </> "$pkgid"),
+                          then "$topdir/../doc/html/libraries/$pkg"
+                          else (myDocdir </> "$pkg"),
           htmldir   = toPathTemplate "$docdir"
       }
 
