diff --git a/compiler/main/Packages.hs b/compiler/main/Packages.hs
index 949cc0f231..408794e078 100644
--- a/compiler/main/Packages.hs
+++ b/compiler/main/Packages.hs
@@ -1339,7 +1339,7 @@ validateDatabase dflags pkg_map1 =
     unusable = directly_ignored `Map.union` unusable_ignored
                                 `Map.union` unusable_broken
                                 `Map.union` unusable_cyclic
-                                `Map.union` unusable_shadowed
+                                -- `Map.union` unusable_shadowed
 
 -- -----------------------------------------------------------------------------
 -- When all the command-line options are in, we can process our package
